### PR TITLE
Ansible EdgeConfig and PlaybookExecution controllers

### PR DESCRIPTION
**Pull request by gciavarrini**

### CODE REVIEW

File: internal/edgeapi/yggdrasil/yggdrasil.go [hunk 1]
```diff
@@ -263,6 +275,44 @@ func (h *Handler) PostDataMessageForDevice(ctx context.Context, params yggdrasil
 
 		h.metrics.IncEdgeDeviceSuccessfulRegistration()
 		return operations.NewPostDataMessageForDeviceOK().WithPayload(&res)
+	case "ansible":
+		ns := h.getNamespace(ctx)
+
+		edgeDevice, err := h.edgeDeviceRepository.Read(ctx, deviceID, ns)
```
Comment (gciavarrini) [inline internal/edgeapi/yggdrasil/yggdrasil.go:281]: @jakub-dzon I need to use the `edgeDeviceRepository`: is there a better way to get it after the split in `backend` and `api`?

**Comment by eloycoto** [comment]
Hi! 

After debugging tooo much this, I'm a bit lost on how this should run:

For me the sysadmin creates the Edgeconfig entry with the following:


```
$ -> export PLAYBOOK=$(base64 -w0 <<'EOF'
---
- name: This is a hello-world example
  tasks:
    - name: Create a file called '/tmp/testfile.txt' with the content 'hello world'.
      copy:
        content: hello worldn
        dest: /tmp/testfile.txt
EOF
)

$ -> echo $PLAYBOOK | base64 -d

$ -> cat <<EOF | kubectl apply -f -
apiVersion: management.project-flotta.io/v1alpha1
kind: EdgeConfig
metadata:
  name: global-config
spec:
  edgePlaybook:
    playbooks:
    - content: ${PLAYBOOK}
EOF
```


So afterwards the device need to have the label:

```
$ -> kubectl label edgedevice fooa config/device-by-config=global-config
```

And this will create the following PlaybookExecutions:

```
$ -> kubectl get playbookexecutions
NAME                 AGE
fooa-global-config   26s
foob-global-config   26s
$ ->
```

So, in this PR, where the playbookexecutions are retrieved from the device[1]:

https://github.com/project-flotta/flotta-operator/blob/0d9002ec32fa7b3aca07dc2ab7e96b0caa75b1ca/internal/edgeapi/yggdrasil/yggdrasil.go#L277-L307


But we expect the ansible config on the getConfig in the device:

https://github.com/project-flotta/flotta-device-worker/blob/e02e796ca1f914ebc6f22cbdba23130f283c9f41/internal/ansible/ansible_manager.go#L101-L120

But we're not setting it in any place. On the other hand, DeviceConfiguration
message is only one string, where maybe should be  []string

```
curl -s \
  --cacert /etc/pki/consumer/ca.pem
  --cert /etc/pki/consumer/cert.pem
  --key /etc/pki/consumer/key.pem
  -H "Content-Type: application/json"
  https://project-flotta.io:8043/api/flotta-management/v1/data/fooa/in  | jq .

{
  "content": {
    "configuration": {
      "heartbeat": {
        "hardware_profile": {},
        "period_seconds": 60
      },
      "metrics": {
        "receiver": {
          "request_num_samples": 30000,
          "timeout_seconds": 10
        }
      },
      "mounts": []
    },
    "device_id": "fooa",
    "version": "75426",
    "workloads": [
      {
        "name": "powertop",
        "namespace": "default",
        "specification": "containers:\n- image: quay.io/eloycoto0/powertop:latest\n  name: powertop\n  resources: {}\n  securityContext:\n    privileged: true\n"
      }
    ]
  },
  "directive": "device",
  "message_id": "d84ca49b-e3e2-410d-8c8a-684be04e50e9",
  "sent": "2022-08-25T16:34:16.676+02:00",
  "type": "data",
  "version": 1
}
```

So, the main issue is:


- If we use the v1/data/$deviceID/in endpoint, that it's what we use for normal
  configuration, we are not setting the config correctly at all in this PR. For
  me this is where this should belongs to. Mainly to have a single heartbeat
  message

- I guess that using PostDataMessageForDevice(data/deviceId/out) function with
  ansible, its only to report status of a device and report in there? But
  Edgedevice has status too :\

So, we need to take a decission here and validate what we want to achieve.


[1]

```

$ -> cat payload.json | jq .
{
  "content": {},
  "directive": "ansible",
  "message_id": "test",
  "sent": "2021-11-21T14:45:25.271+02:00",
  "type": "data",
  "version": 1
}
$ -> curl -s   --cacert /etc/pki/consumer/ca.pem   --cert /etc/pki/consumer/cert.pem   --key /etc/pki/consumer/key.pem   -H "Content-Type: application/json" -d@payload.json  https://project-flotta.io:8043/api/flotta-management/v1/data/fooa/out  | jq .
{
  "content": "---\n- name: This is a hello-world example\n  tasks:\n    - name: Create a file called '/tmp/testfile.txt' with the content 'hello world'.\n      copy:\n        content: hello worldn\n        dest: /tmp/testfile.txt\n",
  "directive": "ansible",
  "message_id": "4bcacb20-746e-454b-ab0a-fc811df7dbba",
  "metadata": {
    "ansible-playbook": "true"
  },
  "sent": "2022-08-25T17:30:09.029+02:00",
  "type": "data",
  "version": 1
}
$ ->
```

Regards

**Comment by gciavarrini** [comment]
>For me the sysadmin creates the Edgeconfig entry with the following: (......)

Correct


> So afterwards the device need to have the label:
> 
> $ -> kubectl label edgedevice fooa config/device-by-config=global-config

Correct 


> And this will create the following PlaybookExecutions:
> 
> $ -> kubectl get playbookexecutions
> NAME                 AGE
> fooa-global-config   26s
> foob-global-config   26s
> $ ->

Yes, a playbookexecution for each labelled device (in this example there are two devices `fooa` and `foob`)

> 
> So, in this PR, where the playbookexecutions are retrieved from the device[1]:
> 
> [flotta-operator/internal/edgeapi/yggdrasil/yggdrasil.go](https://github.com/project-flotta/flotta-operator/blob/0d9002ec32fa7b3aca07dc2ab7e96b0caa75b1ca/internal/edgeapi/yggdrasil/yggdrasil.go#L277-L307)
> [....]

> But we expect the ansible config on the getConfig in the device:
> 
> https://github.com/project-flotta/flotta-device-worker/blob/e02e796ca1f914ebc6f22cbdba23130f283c9f41/internal/ansible/ansible_manager.go#L101-L120
>
> But we're not setting it in any place. On the other hand, DeviceConfiguration

That's true. 

The flotta device PR (https://github.com/project-flotta/flotta-device-worker/pull/69) has been created/merged before the definition of EdgeConfig and PlaybookExecution CRDs. 
The idea at the time was to send a plain ansible playbook on the the deviceconfiguration message (https://github.com/project-flotta/flotta-device-worker/blob/e02e796ca1f914ebc6f22cbdba23130f283c9f41/internal/ansible/ansible_manager.go#L113).
Whenever the device receives a message with ` Metadata == "ansible-playbook" ` then the playbook execution is triggered.

Need to update the device worker to use the CRs. 

> message is only one string, where maybe should be []string

Can you clarify this statement?
At the moment the message is create here https://github.com/project-flotta/flotta-operator/blob/0d9002ec32fa7b3aca07dc2ab7e96b0caa75b1ca/internal/edgeapi/yggdrasil/yggdrasil.go#L291-L302 and a `PlaybookExecution` can contain only one playbook. 
(Please note that, at the moment, the code can't handle multiple PlaybookExecution CRs for each device. Need to fix this :( )


> So, the main issue is:
> 
>     If we use the v1/data/$deviceID/in endpoint, that it's what we use for normal
>     configuration, we are not setting the config correctly at all in this PR. For
>     me this is where this should belongs to. Mainly to have a single heartbeat
>     message
> 
>     I guess that using PostDataMessageForDevice(data/deviceId/out) function with
>     ansible, its only to report status of a device and report in there? But
>     Edgedevice has status too :\
> 
> So, we need to take a decission here and validate what we want to achieve.

I'd like to discuss it with you, if you don't mind :)

File: cmd/httpapi/main.go [hunk 2]
```diff
@@ -191,7 +191,13 @@ func main() {
 	mux.Handle("/metrics", promhttp.HandlerFor(crmetrics.Registry, promhttp.HandlerOpts{}))
 	mux.HandleFunc("/healthz", httpOK)
 	mux.HandleFunc("/readyz", httpOK)
-	logger.Fatal(http.ListenAndServe(Config.MetricsAddr, mux))
```
Comment (gciavarrini) [inline cmd/httpapi/main.go:187]: @eloycoto  Shouldn't this change be part of a separate PR?

File: controllers/edgeconfig_controller.go [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+	Scheme                      *runtime.Scheme
+	EdgeConfigRepository        edgeconfig.Repository
+	EdgeDeviceRepository        edgedevice.Repository
+	PlaybookExecutionRepository playbookexecution.Repository
+	Concurrency                 uint
+	MaxConcurrentReconciles     int
+	ExecuteConcurrent           func(context.Context, uint, ConcurrentFunc, []v1alpha1.EdgeDevice) []error
+}
+
+var logger = log.FromContext(context.TODO())
```
Comment (eloycoto) [inline controllers/edgeconfig_controller.go:52]: is this global logger correct here? I mean, It'll be updated on line 68.

File: controllers/edgeconfig_controller.go [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+
+	if edgeConfig.DeletionTimestamp != nil {
+		return ctrl.Result{}, nil
+	}
+
+	edgeDevices, err := r.EdgeDeviceRepository.ListForEdgeConfig(ctx, edgeConfig.Name, edgeConfig.Namespace)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return ctrl.Result{Requeue: true}, err
+		}
```
Comment (eloycoto) [inline controllers/edgeconfig_controller.go:87]: if no edgedevices should return here, no?

File: controllers/edgeconfig_controller.go [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+
+		playbookExec.Status = playbookExecutionStatus
+		playbookExecutions = append(playbookExecutions, playbookExec)
+	}
+
+	return playbookExecutions
+}
+
+func (r *EdgeConfigReconciler) addPlaybookExecutionToDevices(ctx context.Context, edgeConfig *v1alpha1.EdgeConfig, edgeDevices []v1alpha1.EdgeDevice) error {
+	playbookExecutionBases := createPlaybookExecutions(edgeConfig)
```
Comment (eloycoto) [inline controllers/edgeconfig_controller.go:145]: I'll return an error if len(edgeDevices) is 0 .

File: controllers/edgeconfig_controller.go [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+						continue
+					}
+				}
+			}
+		}
+		return errs
+	}
+
+	errs := r.executeConcurrent(ctx, f, edgeDevices)
+	if len(errs) != 0 {
```
Comment (eloycoto) [inline controllers/edgeconfig_controller.go:193]: I would use here go multiple-error as we used on device-worker :-)

File: controllers/edgeconfig_controller.go [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+		errs = r.ExecuteConcurrent(ctx, r.Concurrency, f, edgeDevices)
+	}
+	return errs
+}
+
+func ExecuteEdgeConfigConcurrent(ctx context.Context, concurrency uint, f ConcurrentFunc, edgeDevices []v1alpha1.EdgeDevice) []error {
+	if len(edgeDevices) == 0 || concurrency == 0 {
+		return nil
+	}
+	inputs := splitEdgeDevices(edgeDevices, concurrency)
```
Comment (eloycoto) [inline controllers/edgeconfig_controller.go:222]: this is pretty cool :D

File: controllers/edgeconfig_controller_test.go [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+				Scheme:                      k8sManager.GetScheme(),
+				EdgeConfigRepository:        edgeConfigRepoMock,
+				EdgeDeviceRepository:        edgeDeviceRepoMock,
+				PlaybookExecutionRepository: playbookExecRepoMock,
+				Concurrency:                 1,
+				MaxConcurrentReconciles:     1,
+			}
+		})
+
+		// edgePlaybookSpec := v1alpha1.EdgePlaybookSpec{
```
Comment (eloycoto) [inline controllers/edgeconfig_controller_test.go:98]: remove these lines.

File: controllers/playbookexecution_controller_test.go [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+	var (
+		playbookExecutionReconciler *controllers.PlaybookExecutionReconciler
+		err                         error
+		cancelContext               context.CancelFunc
+		signalContext               context.Context
+		req                         ctrl.Request
+
+		playbookExecutionRepoMock *playbookexecution.MockRepository
+		edgeDeviceRepoMock        *edgedevice.MockRepository
+		// playbookExecRepoMock      *playbookexecution.MockRepository
```
Comment (eloycoto) [inline controllers/playbookexecution_controller_test.go:33]: remove these comments.

File: controllers/playbookexecution_controller_test.go [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+			// when
+			res, err := playbookExecutionReconciler.Reconcile(context.TODO(), req)
+
+			// then
+			Expect(err).NotTo(HaveOccurred())
+			Expect(res).To(Equal(reconcile.Result{Requeue: false, RequeueAfter: 0}))
+		})
+	})
+
+	// Context("edgeDevice selection", func() {
```
Comment (eloycoto) [inline controllers/playbookexecution_controller_test.go:127]: adn these ones.

File: internal/edgeapi/backend/k8s/heartbeat-updater.go [hunk 11]
```diff
@@ -78,3 +123,25 @@ func updateDeploymentStatuses(oldWorkloads []v1alpha1.Workload, workloads []*mod
...
+func updatePlaybookExecutionStatuses(oldPlaybookExecution []v1alpha1.PlaybookExec, playbookExecutions []*models.PlaybookExecutionStatus) []v1alpha1.PlaybookExec {
+	playbookExecutionMap := make(map[string]v1alpha1.PlaybookExec)
+	for _, peStatus := range oldPlaybookExecution {
+		playbookExecutionMap[peStatus.Name] = peStatus
+	}
+	for _, status := range playbookExecutions {
+		if newPlayExec, ok := playbookExecutionMap[status.Name]; ok {
+			if string(newPlayExec.PlaybookExecutionStatus.Conditions[len(newPlayExec.PlaybookExecutionStatus.Conditions)-1].Type) != status.Status {
+				newPlayExec.PlaybookExecutionStatus.Conditions[len(newPlayExec.PlaybookExecutionStatus.Conditions)-1].Status = v1.ConditionFalse
+				now := v1.Now()
```
Comment (eloycoto) [inline internal/edgeapi/backend/k8s/heartbeat-updater.go:136]: I'm worried that this will match at any momemtn

File: controllers/edgeconfig_controller.go (hunk 4 above)
Comment (gciavarrini) [reply controllers/edgeconfig_controller.go:87]: Do you mean I should add a `return ctrl.Result{}, err` or  `return ctrl.Result{true}, err`  in line 88 ?

File: internal/edgeapi/backend/k8s/heartbeat-updater.go (hunk 11 above)
Comment (gciavarrini) [reply internal/edgeapi/backend/k8s/heartbeat-updater.go:136]: Need more info here. What do you mean with "match"?
I'm creating a new Condition and I must set the `LastTransitionTime` that, in my opinion, is `Now()`. What I'm missing?

**Comment by pkliczewski** [comment]
@gciavarrini @eloycoto is there anything else we would like to do to finalize this PR?

File: controllers/edgeconfig_controller.go (hunk 7 above)
Comment (gciavarrini) [reply controllers/edgeconfig_controller.go:226]: not my idea/code :D

File: controllers/edgeconfig_controller.go (hunk 4 above)
Comment (gciavarrini) [reply controllers/edgeconfig_controller.go:85]: Added. 
I also included a len check on the devices slice.

File: controllers/edgeconfig_controller.go (hunk 5 above)
Comment (gciavarrini) [reply controllers/edgeconfig_controller.go:149]: Added in `Reconcile` method, not necessary anymore here (but thanks for the catch!)

File: controllers/edgeconfig_controller.go (hunk 6 above)
Comment (gciavarrini) [reply controllers/edgeconfig_controller.go:197]: If you don't have a strong opinion I'd prefer to use array of errors for the moment and create a separate PR (if needed) to address this suggestion.

This code uses methods already defined and used by other controllers. That change will impact several files, so I'd like to do not increase the complexity of this PR (that already contains a lot of changes).

**Comment by eloycoto** [comment]
I'm ok to merge, I think that all is addressed already
